### PR TITLE
[Merged by Bors] - feat: port data.{nat, int}.cast.defs

### PR DESCRIPTION
**Pull request by j-loreaux**
mathlib3 SHA: bb168510ef455e9280a152e7f31673cabd3d7496

Porting notes for `Mathlib.Data.Nat.Cast.Defs`

1. I have moved `AddMonoidWithOne` and related lemmas from `Mathlib.Algebra.GroupWithZero.Defs` into this file. All but 3 of these were in this file in mathlib3, so this is the correct place. The extra 3 are new lemmas in Mathlib4, and I made a note to this effect.
2. mathport marked some translations as dubious because they used `AddMonoidWithOne` instead of `NatCast` because the latter didn't exist until this PR, so I just fixed them to use `NatCast` and `#align`ed (with no `ₓ`).
3. I need someone (e.g., @gebner) to rewrite the library note about coercions because I don't know enough.
4. I reimplemented `Nat.binCast` from scratch because we don't have `Nat.binaryRec`. The proof of `Nat.binCast_eq` can likely be improved quite a bit.
5. I marked `Nat.cast_bit0` and `Nat.cast_bit1` as deprecated since `bit0` and `bit1` are deprecated.
6. I added an instance of `OfNat R n` for `n : ℕ` when `R` has an `AddMonoidWithOne` instance. I think we want this in order to talk about literals as terms in `R` (e.g., `37 : R`). This instance does not appear in mathlib3 and is new.
7. ~~I kept the (lack of) `simp` attributes from mathlib3 for certain lemmas which were already ported to mathlib4. This caused some breakage which I fixed. I thought it was better to be consistent with the mathlib3 `simp` set.~~ Per Gabriel's suggestion, I have kept these as `simp` lemmas.

Porting notes for `Mathlib.Data.Int.Cast.Defs`

1. This file already existed in Mathlib4, but it's contents corresponded to things in mathlib3's `data.int.cast.basic`, so I have moved the previous contents to `Mathlib.Data.Int.Cast.Basic`, hence the reason for the weird diff.
2. I have moved `AddGroupWithOne` and the rest of that material out of `Mathlib.Algebra.GroupWithZero.Defs` so these files are now independent and `Mathlib.Data.Int.Cast.Defs` I think is a faithful and complete translation of the mathlib3 file.

### CODE REVIEW

**Comment by digama0** [comment]
> This file has been modified in a more recent mathlib3 commit, namely #17477: bb168510ef455e9280a152e7f31673cabd3d7496, but the changes were minor and I have tried to incorporate them here.

In that case you should put this SHA at the top of the PR message (unless you are not sure you were effective in incorporating the changes?).

**Comment by j-loreaux** [comment]
Okay, I've changed it. It was only a rename of a single lemma.

File: Mathlib/Data/Nat/Cast/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+-- see note [coercion into rings]
+instance [NatCast R] : CoeTail ℕ R where coe := Nat.cast
+
+-- see note [coercion into rings]
+instance [NatCast R] : CoeHTCT ℕ R where coe := Nat.cast
+
+-- I *think* we want this instance? Otherwise we can't write `2 : R` (see `Nat.cast_two` below)
+-- this was not in the Lean 3 code
+instance [NatCast R] (n : ℕ) : OfNat R n where ofNat := Nat.cast n
```
Comment (semorrison) [inline Mathlib/Data/Nat/Cast/Defs.lean:57]: Perhaps provide a doc-string on this instance foreshadowing the simp lemma `Nat.cast_ofNat` below, to explain what the simp norm form is.

Comment (j-loreaux) [reply Mathlib/Data/Nat/Cast/Defs.lean]: Thanks, you made me realize that the extra bits at the bottom should have been moved up, and they included this `OfNat` instance.

I *think* if I understood Mario correctly yesterday, that `Nat.cast` is the `simp` normal form, but for literals you always have to have `OfNat.ofNat`, so I don't think that `Nat.cast_ofNat` is dictating the `simp` normal form, but maybe I have terribly misunderstood something.

**Comment by gebner** [comment]
> I kept the (lack of) simp attributes from mathlib3 for certain lemmas which were already ported to mathlib4. This caused some breakage which I fixed. I thought it was better to be consistent with the mathlib3 simp set.

I think we should keep the `@[simp]` here for now.  It's different in mathlib3 because of a recent PR: https://github.com/leanprover-community/mathlib/pull/17048

However, that PR is going to be very ugly to port to Lean 4, since we no longer have a `coe` constant.

**Comment by j-loreaux** [comment]
I have reinstated the `simp` attribute on these lemmas.

**Comment by j-loreaux** [comment]
Note in case you already looked at this PR and didn't read the updated description: I have added a port of `data.int.cast.defs` into this PR.

**Comment by semorrison** [comment]
bors merge

**Comment by gebner** [comment]
bors merge

**Pull request closed without merging**